wagtail/admin/templates/wagtailadmin/shared/headers/header_with_avatar.html

```diff
@@ -0,0 +1,17 @@
+{% comment "text/markdown" %}
+
+    Variables accepted by this template:
+
+    - `title` - Displayed as `h1`
+    - `subtitle` - Within the `h1` tag but smaller
+    - `avatar` - if present, display an 'avatar' in place of icon. This is the URL to be used as the img src for avatar
+{% endcomment %}
+<header class="w-header w-header--with-avatar">
+    {% if avatar %}<div class="avatar"><img src="{{ avatar }}" alt=""/></div>{% endif %}
+    <div>
+        <h1 id="header-title" class="w-header__title">
+            {{ title }}
+            {% if subtitle %} <span class="w-header__subtitle">{{ subtitle }}</span>{% endif %}
+        </h1>
+    </div>
+</header>
```

Comment:
There must be a better way to be able to use the shared header template instead of adding a new header template.

I know the existing template may be doing a bit much but one of the key outputs of the UX unification project was the reduction of about 4+ header templates into just two (if you ignore the slim header new ones).

The addition of the avatar into the shared header template was intentional, so not sure why we should pull it out.

Reply:
The header.html markup isn't quite built for this design and the home header is unique to that page so why should header.html support that one instance?
 
I do agree its nice to have one reusable template that we can use for the headers but at some point having templates with more specific purposes and less logic/css entanglement can be easier to understand modify and reuse. If we do use the current header.html for this we end up throwing more weird negative margins/crazy padding or positioning to make everything align.

Reply:
I agree but this is the direction we went towards intentionally so we could avoid more templates.
I see your perspective, which is why it was such a pain to get it working as well as it is now with the UX unification project.

Ideally we should not need all the messy padding/margins etc - but at the moment that is the only real way to make the header work with the sidebar toggle on mobile + support the modal variants.

see https://github.com/wagtail/wagtail/issues/8906 for future suggested improvements here
see https://github.com/wagtail/wagtail/issues/8907 to adopt the same shared header component/template in other parts.

I think it is worth another go to try to get the design alignment done with the existing HTML structure personally, I am happy to help with some suggested styling if that would be useful.

The core idea behind this approach is to align with one template structure for all headers (excluding slim headers and new modals), then I think we will be able to revisit the DOM structure's suitability as a separate change on top (as per issues above).

While styling wrangling can be a bit frustrating, the alternative is a propagation of small variations in the HTML structure.

Reply:
I can make the styling work with this current markup however it feels like a wasted effort when there are simpler ways to achieve it. 

I'll revise this for now to use the existing so we can get this work in but we will definitely have to restyle it yet again when the header's floats/extra markup is removed in favor of flexbox (can't wait for that 👀)

Reply:
thanks  - and yes there is an issue for that (removal of floats)
https://github.com/wagtail/wagtail/issues/8126

Reply:
@lb- if the goal is to avoid propagation of different header templates, would you be more keen on the dashboard having its header markup directly inside its own template?

Personally I think the furthest "dashboard-only" markup is from our reusable header component, the better. There is no intention to support displaying avatars in headers anywhere but this type of "vanity" screen, and similarly using the very large font size is only something that should be used in such one-off scenarios.

So to me the added complexity for the one reusable template to support this style variation as well doesn’t feel warranted – that header component already has plenty enough variations it needs to support.

Reply:
ok, that kind of goes against what I understood we were trying to do but fully understand this reasoning.

How about this approach below, basically just provide a supported way to override everything around, including, the `h1` tag.

This should give us enough flexibility and also solve for future planned work for modeladmin support. We risk adding complexity plus allow for a missing `h1` element (accessibility).

This also means removing the `avatar` thing and the `glyph` classes probably completely from the shared header template AND styles.

### Option 1 - use template inheritance and add a `h1` block (easiest)

* This will help us towards resolving the modeladmin's ability to provide a `h1` block override - see https://github.com/wagtail/wagtail/issues/8907
* This provides a way to just make this core title area quite custom but will just mean we must include the base `h1` markup (id & class) , but everything else inside can be custom.

```html
<header class="{% classnames "w-header" classname merged|yesno:"w-header--merged," search_form|yesno:"w-header--hasform," %}">
    {% block breadcrumb %}{% endblock %}
    <div class="row">
        <div class="left">
            <div class="col">
                {% block h1 %}
                    <h1 class="w-header__title" id="header-title">
                        {% if icon %}{% icon class_name="w-header__glyph" name=icon %}{% endif %}
                        {{ title }}{% if subtitle %} <span class="w-header__subtitle">{{ subtitle }}</span>{% endif %}
                    </h1>
                {% endblock %}
                {% if description %}<div class="w-header__description">{{ description }}</div>{% endif %}
            </div>
```



### Option 2 - use template children approach with a new arg

* This will also help us towards resolving the modeladmin's ability to provide a `h1` block override - but we may need to make some tweaks to ensure we can use it for the modeladmin header work.
* Similar to the above but we use something like `h1` or even `children` to be able to pass in a fully custom bit of content in place of the built in `h1` title use?
* Not sure if this should also envelop the `description` or even the entire first `col`

```html
<header class="{% classnames "w-header" classname merged|yesno:"w-header--merged," search_form|yesno:"w-header--hasform," %}">
    {% block breadcrumb %}{% endblock %}
    <div class="row">
        <div class="left">
            <div class="col">
                {% if children %}
                    {{ children }}
                {% else %}     
                    <h1 class="w-header__title" id="header-title">
                        {% if icon %}{% icon class_name="w-header__glyph" name=icon %}{% endif %}
                        {{ title }}{% if subtitle %} <span class="w-header__subtitle">{{ subtitle }}</span>{% endif %}
                    </h1>
                    {% if description %}<div class="w-header__description">{{ description }}</div>{% endif %}
                {% endif %}
            </div>
```

Reply:
Sorry for pushing back so hard on this - just really trying to reconcile the goals of a design that is unique for this one page. Along with the desire to have templates (eventually template components) that are reasonably consistent and hopefully more maintainable over the long term.

But 100% happy to go with your call here @thibaudcolas  - just want to make sure we build on, not pull a part, the work done with the UX unification project.

Thanks for being flexible @stevedya